drivers/kscan/kscan_gpio.c

```diff
@@ -0,0 +1,495 @@
+/*
+ * Copyright (c) 2020 Peter Johanson <peter@peterjohanson.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gpio_kscan
+
+#include <device.h>
+#include <drivers/kscan.h>
+#include <drivers/gpio.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(kscan_gpio, CONFIG_KSCAN_LOG_LEVEL);
+
+#define MATRIX_NODE_ID DT_DRV_INST(0)
+#define MATRIX_ROWS DT_PROP_LEN(MATRIX_NODE_ID, row_gpios)
+#define MATRIX_COLS DT_PROP_LEN(MATRIX_NODE_ID, col_gpios)
+
+#if DT_ENUM_IDX(MATRIX_NODE_ID, diode_direction) == 0
+#define SCAN_ROW_TO_COLUMN
+#else
+#define SCAN_COL_TO_ROW
+#endif
+
+struct kscan_gpio_item_config {
+	char *label;
+	gpio_pin_t pin;
+	gpio_flags_t flags;
+};
+
+struct kscan_gpio_config {
+	u8_t debounce_period;
+	struct kscan_gpio_item_config rows[MATRIX_ROWS];
+	struct kscan_gpio_item_config cols[MATRIX_COLS];
+};
+
+struct kscan_gpio_data {
+	kscan_callback_t callback;
+#if DT_PROP(MATRIX_NODE_ID, debounce_id) == 0
+	struct k_work work;
+#else
+	struct k_delayed_work work;
+#endif
+	bool matrix_state[MATRIX_ROWS][MATRIX_COLS];
```

Comment:
Before the first interrupt matrix_state contains uninitialized values. As a result, on the very first keypress we get long sequence of callbacks for every key.
kscan_gpio_init() should call dummy kscan_gpio_read() with empty callback function.